src/types/FindSelected.ts

```diff
@@ -0,0 +1,12 @@
+import type { SelectPattern } from './Pattern';
+import type { ValueOf } from './helpers';
+
+export type FindSelected<a, b> = b extends SelectPattern<infer Key>
+  ? { [k in Key]: a }
+  : [a, b] extends [(infer aa)[], [infer p]]
+  ? FindSelected<aa, p> extends infer selected
```

Comment:
In this PR, I'm using top level type recursion and variadic tuples, which are only supported since TS 4.1 and TS 4 respectively. It makes this a breaking change because people using earlier versions of TS will have to upgrade.